File: Data Structures & Algorithms/validate-parentheses/submission-0.go
func isValid(s string) bool {
	if len(s)%2 != 0 {
		return false
	}

	m := map[byte]byte{'{': '}', '[': ']', '(': ')'}
	stack := make([]byte, 0, len(s)/2)

	for i := range s {
		if val, ok := m[s[i]]; ok {
			stack = append(stack, val)
		} else {
			if len(stack) == 0 || stack[len(stack)-1] != s[i] {
				return false
			}
			stack = stack[:len(stack)-1]
		}
	}

	return len(stack) == 0
}

File: Data Structures & Algorithms/validate-parentheses/submission-0.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let map = HashMap::from([('(', ')'), ('[', ']'), ('{', '}')]);

        let mut stack = Vec::<char>::new();

        for ch in s.chars() {
            if map.contains_key(&ch) {
                stack.push(*map.get(&ch).unwrap());
            } else if stack.pop().unwrap_or('.') != ch {
                return false;
            }
        }

        stack.is_empty()
    }
}
